Makefile

```
@@ -106,58 +106,68 @@ $(K0S_GO_BUILD_CACHE):
 	rm -rf 'static/manifests/$(gen_output_dir)/CustomResourceDefinition'
-	rm -f -- '$(dir $@)'zz_*.go
-	CGO_ENABLED=0 $(GO) install sigs.k8s.io/controller-tools/cmd/controller-gen@v$(controller-gen_version)
-	$(GO_ENV) controller-gen \
-	  crd \
-	  paths="./$(dir $@)..." \
-	  output:crd:artifacts:config=./static/manifests/$(gen_output_dir)/CustomResourceDefinition \
-	  object:headerFile=hack/tools/boilerplate.go.txt
+	mkdir -p 'static/manifests/$(gen_output_dir)'
+	gendir="$$(mktemp -d .controller-gen.XXXXXX.tmp)" \
+	  && trap "rm -rf -- $$gendir" INT EXIT \
+	  && CGO_ENABLED=0 $(GO) run sigs.k8s.io/controller-tools/cmd/controller-gen@v$(controller-gen_version) \
+	    paths="./$(dir $@)..." \
+	    object:headerFile=hack/tools/boilerplate.go.txt output:object:dir="$$gendir" \
+	    crd output:crd:dir='static/manifests/$(gen_output_dir)/CustomResourceDefinition' \
+	  && mv -f -- "$$gendir"/zz_generated.deepcopy.go '$(dir $@).'
 	touch -- '$@'
 
 clientset_input_dirs := pkg/apis/autopilot/v1beta2 pkg/apis/k0s/v1beta1
 codegen_targets += pkg/client/clientset/.client-gen.stamp
 pkg/client/clientset/.client-gen.stamp: $(shell find $(clientset_input_dirs) -type f -name \*.go -not -name \*_test.go -not -name zz_\*)
 pkg/client/clientset/.client-gen.stamp: .k0sbuild.docker-image.k0s hack/tools/boilerplate.go.txt embedded-bins/Makefile.variables
-	rm -rf -- '$(dir $@)'
-	CGO_ENABLED=0 $(GO) run k8s.io/code-generator/cmd/client-gen@v$(kubernetes_version:1.%=0.%) \
-	  --go-header-file=hack/tools/boilerplate.go.txt \
-	  --input-base='' \
-	  --input=$(subst $(space),$(comma),$(clientset_input_dirs:%=github.com/k0sproject/k0s/%)) \
-	  --output-package=github.com/k0sproject/k0s/$(patsubst %/,%,$(dir $(patsubst %/,%,$(dir $@)))) \
-	  --clientset-name=$(notdir $(patsubst %/,%,$(dir $@))) \
-	  --output-base=. \
-	  --trim-path-prefix=github.com/k0sproject/k0s
+	gendir="$$(mktemp -d .client-gen.XXXXXX.tmp)" \
+	  && trap "rm -rf -- $$gendir" INT EXIT \
+	  && CGO_ENABLED=0 $(GO) run k8s.io/code-generator/cmd/client-gen@v$(kubernetes_version:1.%=0.%) \
+	    --go-header-file=hack/tools/boilerplate.go.txt \
+	    --input-base='' \
+	    --input=$(subst $(space),$(comma),$(clientset_input_dirs:%=github.com/k0sproject/k0s/%)) \
+	    --output-package=github.com/k0sproject/k0s/pkg/client \
+	    --clientset-name=clientset \
+	    --output-base="$$gendir/out" \
+	    --trim-path-prefix=github.com/k0sproject/k0s \
+	  && { [ ! -e pkg/client/clientset ] || mv -- pkg/client/clientset "$$gendir/old"; } \
+	  && mv -f -- "$$gendir/out/github.com/k0sproject/k0s/pkg/client/clientset" pkg/client/.
 	touch -- '$@'
 
 codegen_targets += static/zz_generated_assets.go
-static_asset_dirs := static/manifests static/misc
+static/zz_generated_assets.go: $(controllergen_targets) # to generate the CRDs into static/manifests/*/CustomResourceDefinition
+static/zz_generated_assets.go: $(shell find static/manifests/calico static/manifests/windows static/misc -type f)
```

Not sure how many times or when the dependency is evaluated, but this ensure that we only executes the `find` command only once.

```suggestion
static_asset_files := $(shell find static/manifests/calico static/manifests/windows static/misc -type f)
static/zz_generated_assets.go: $(controllergen_targets) # to generate the CRDs into static/manifests/*/CustomResourceDefinition
static/zz_generated_assets.go: $(static_asset_files)
```

---

Isn't that done only once? I mean, it's the requisites. I can't imagine that make won't cache those? It would probably break everything, because they're used to construct the build graph, meaning that the build plan might change in-flight? We already have some other places with that pattern.